File: taurus-protect-sdk-typescript/src/mappers/webhook-call.ts
/**
 * Webhook call mapper functions for converting OpenAPI DTOs to domain models.
 */

import type { WebhookCall } from '../models/webhook-call';
import type { TgvalidatordWebhookCall } from '../internal/openapi/models';
import { safeDate, safeMap, safeString } from './base';

/**
 * Maps a webhook call DTO to a WebhookCall domain model.
 *
 * @param dto - The OpenAPI webhook call DTO
 * @returns The domain model webhook call, or undefined if dto is null/invalid
 */
export function webhookCallFromDto(dto: TgvalidatordWebhookCall | unknown): WebhookCall | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    id: safeString(d.id),
    eventId: safeString(d.eventId ?? d.event_id),
    webhookId: safeString(d.webhookId ?? d.webhook_id),
    payload: safeString(d.payload),
    status: safeString(d.status),
    statusMessage: safeString(d.statusMessage ?? d.status_message),
    attempts: safeString(d.attempts),
    updatedAt: safeDate(d.updatedAt ?? d.updated_at),
    createdAt: safeDate(d.createdAt ?? d.created_at),
  };
}

/**
 * Maps an array of webhook call DTOs to WebhookCall domain models.
 *
 * @param dtos - The list of OpenAPI webhook call DTOs
 * @returns The list of domain model webhook calls
 */
export function webhookCallsFromDto(
  dtos: TgvalidatordWebhookCall[] | unknown[] | null | undefined
): WebhookCall[] {
  return safeMap(dtos, webhookCallFromDto);
}

File: taurus-protect-sdk-go/pkg/protect/mapper/webhook_call.go
package mapper

import (
	"strconv"

	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/openapi"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// WebhookCallFromDTO converts an OpenAPI WebhookCall to a domain WebhookCall.
func WebhookCallFromDTO(dto *openapi.TgvalidatordWebhookCall) *model.WebhookCall {
	if dto == nil {
		return nil
	}

	call := &model.WebhookCall{
		ID:            safeString(dto.Id),
		EventID:       safeString(dto.EventId),
		WebhookID:     safeString(dto.WebhookId),
		Payload:       safeString(dto.Payload),
		Status:        safeString(dto.Status),
		StatusMessage: safeString(dto.StatusMessage),
	}

	// Parse attempts count (comes as string from API)
	if dto.Attempts != nil {
		if attempts, err := strconv.ParseInt(*dto.Attempts, 10, 64); err == nil {
			call.Attempts = attempts
		}
	}

	// Convert timestamps
	if dto.CreatedAt != nil {
		call.CreatedAt = *dto.CreatedAt
	}
	if dto.UpdatedAt != nil {
		call.UpdatedAt = *dto.UpdatedAt
	}

	return call
}

// WebhookCallsFromDTO converts a slice of OpenAPI WebhookCall to domain WebhookCalls.
func WebhookCallsFromDTO(dtos []openapi.TgvalidatordWebhookCall) []*model.WebhookCall {
	if dtos == nil {
		return nil
	}
	calls := make([]*model.WebhookCall, len(dtos))
	for i := range dtos {
		calls[i] = WebhookCallFromDTO(&dtos[i])
	}
	return calls
}
